### ISSUE TITLE
Incorrect serialization of nested filtering with AND operator

### ISSUE THREAD

**Issue by swyrazik** [crnk-project/crnk-framework#680]
When building [complex filter specifications](http://www.crnk.io/releases/stable/documentation/#_nested_filtering) with the QuerySpec/FilterSpec API and using them to execute queries with the CrnkClient, filters nested under an AND operator are serialized incorrectly. The issue is especially evident and damaging when the nested filters refer to different attributes of the same relation, where former filters get overwritten by latter filters. There is no issue when nesting filters under an OR operator.

I have narrowed the problem down to the [JsonFilterSpecMapper](https://github.com/crnk-project/crnk-framework/blob/master/crnk-core/src/main/java/io/crnk/core/queryspec/internal/JsonFilterSpecMapper.java) class and, more specifically, its serialize method.

The following two tests demonstrate what is wrong with the AND operator (and that the OR operator works fine):

```java
public class JsonFilterSpecMapperTest {

    ...

    @Test
    public void filterTwoAttributesOfSameRelationUsingAnd() {
        final FilterSpec filterSpec = FilterSpec.and(
                new FilterSpec(PathSpec.of("project.name"), FilterOperator.EQ, "test"),
                new FilterSpec(PathSpec.of("project.description"), FilterOperator.EQ, "test test")
        );

        final List<FilterSpec> serializedFilterSpecs = Collections.singletonList(filterSpec);
        final JsonNode jsonNode = mapper.serialize(serializedFilterSpecs);

        Assert.assertEquals("{\"AND\":[{\"project\":{\"name\":\"test\"}},{\"project\":{\"description\":\"test test\"}}]}", jsonNode.toString());
        // actual: {"AND":{"project":{"description":"test test"}}}
        // the filter for project.name disappeared
    }

    @Test
    public void filterTwoAttributesOfSameRelationUsingOr() {
        final FilterSpec filterSpec = FilterSpec.or(
                new FilterSpec(PathSpec.of("project.name"), FilterOperator.EQ, "test"),
                new FilterSpec(PathSpec.of("project.description"), FilterOperator.EQ, "test test")
        );

        final List<FilterSpec> serializedFilterSpecs = Collections.singletonList(filterSpec);
        final JsonNode jsonNode = mapper.serialize(serializedFilterSpecs);

        Assert.assertEquals("{\"OR\":[{\"project\":{\"name\":\"test\"}},{\"project\":{\"description\":\"test test\"}}]}", jsonNode.toString());
    }
}
```

**Comment by remmeier** [comment]
do you see by chance the fix?

**Comment by swyrazik** [comment]
Not currently. I'd have to spend some time on it to figure it out.

I was hoping that you'd be able to see the fix more easily. If not, I'll have to dig in because I need this working for a project.

**Comment by remmeier** [comment]
merged, 3.2 to be released this week, closing the issue. thank you for the fix!
